### ISSUE TITLE
Providing single speaker labels via HDFDataset

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#388]
This issue is somewhat related to #331, but I still want to know if what I have is a wrong approach or a bug.

I have the following dataset:

```python
{ 'class': 'MetaDataset',
  'data_map': {'classes': ('audio', 'classes'), 'data': ('audio', 'data'), 'speaker_label': ('speaker', 'data')},
  'datasets': { 'audio': {[...]},
                'speaker': { 'class': 'HDFDataset',
                             'files': [ '/u/rossenbach/experiments/librispeech_tts/work/tts/speaker_embeddings/SpeakerLabelHDFFromBliss.xmjLCiVxKiEW/output/speaker_labels.hdf']}},
```

This is my data definition:
```python3
extern_data = {[...],
  'speaker_label': {'dim': 251, 'shape': (None,), 'sparse': True}}
```
And the stored data format is `[B,T|1]`

Unfortunately there is still not a way to store the data without a time axis, so I am doing the same as I did for my speaker vectors before, but now for the single label:
```python3
'speaker_label_notime': {'axis': 'T', 'class': 'squeeze', 'from': ['data:speaker_label']},
```

This results in the following assertion fail (the rest of the trace is not relevant):
```python3
File "/work/asr4/rossenbach/sisyphus_work_folders/librispeech_tts_work/returnn/helpers/rossenbach/checkout/CheckoutRETURNN.WE8k4r4LocNz/output/crnn/b1674b45c0449dc0fef7350bd22abaa5a03036e4/returnn/tf/layers/basic.py", line 3642, in get_out_data_from_opts
    line: return ReduceLayer.get_out_data_from_opts(
    locals:
      ReduceLayer = <global> <class 'returnn.tf.layers.basic.ReduceLayer'>
      ReduceLayer.get_out_data_from_opts = <global> <bound method ReduceLayer.get_out_data_from_opts of <class 'returnn.tf.layers.basic.ReduceLayer'>>
  File "/work/asr4/rossenbach/sisyphus_work_folders/librispeech_tts_work/returnn/helpers/rossenbach/checkout/CheckoutRETURNN.WE8k4r4LocNz/output/crnn/b1674b45c0449dc0fef7350bd22abaa5a03036e4/returnn/tf/layers/basic.py", line 3485, in get_out_data_from_opts
    line: assert not x.sparse
    locals:
      x = <local> Data(name='speaker_label', shape=(None,), dtype='int32', sparse=True, dim=251, available_for_inference=False, batch_shape_meta=[B,T|'time:var:extern_data:speaker_label'])
      x.sparse = <local> True
AssertionError
```
Which is caused by the `SqueezeLayer` using the `get_out_data_from_opts` function from the `ReduceLayer which apparently does not support sparse data.

**Comment by albertz** [comment]
The title of this bug doesn't sound like it is related to your actual bug/exception?

Fixing `SqueezeLayer` should not be a problem, right? It is clear that `SqueezeLayer` should work also for sparse data (or whatever data).

Don't mix up things. This bug should not be about multiple things, like MetaDataset, HDFDataset, SqueezeLayer, ReduceLayer. I'm not actually sure which of these you think is wrong here. But the exception you posted is about `SqueezeLayer`, and yes this is a bug (obviously, right?), and this should be simple to fix (right?).

**Comment by JackTemaki** [comment]
This was not intended to be an issue about a bug, but rather the situation in general. You said it is a bug, which I did consider, but could not be sure of (this is why I asked for it). And this is not simple to fix, as it is not clear to me from the code why sparse is not allowed, and I have to write new test cases for the `ReduceLayer` as well.

**Comment by albertz** [comment]
This is about `SqueezeLayer`, not about `ReduceLayer`. As I said, there is no reason why `SqueezeLayer` should not work on sparse data -- it is clear that it should work no matter if the data is sparse or not. This is clear, right? So this is an obvious bug.

But everything else you write is totally unrelated to all of that.

**Comment by albertz** [comment]
I still don't quite understand this issue report. The bug in `SqueezeLayer` is fixed now. If there is sth in `HDFDataset` or `MetaDataset` (which is clearly totally independent from `SqueezeLayer`), please open a separate issue.

**Comment by JackTemaki** [comment]
I think #331 is sufficient already to describe the remaining problem in the HDFDataset and HDF generation in general.
